### PR TITLE
Release v4

### PR DESCRIPTION
**Pull request by kamilmysliwiec**
- async components & providers & factories & middlewares,
- guards feature,
- pipes & filters for microservices & websockets,
- e2e testing utilities, better unit test utils

### CODE REVIEW

File: src/microservices/package.json [hunk 1]
```diff
@@ -10,12 +10,12 @@
   "dependencies": {
     "json-socket": "^0.2.1",
     "iterare": "0.0.8",
-    "redis": "^2.6.5"
+    "redis": "^2.7.1"
```
Comment (d3viant0ne) [inline src/microservices/package.json]: I'd normally agree with you but in this particular case, it's an opt-in feature so very little harm given the last of competing technologies.

https://github.com/kamilmysliwiec/nest/blob/eda5c0c0c1e41ae0b1b353cfaa9a733a01ca53b1/src/microservices/server/server-redis.ts

**Comment by d3viant0ne** [comment]
@kamilmysliwiec - Might not hurt to push an alpha / beta off the v4 branch to get people beating on it.

**Comment by KatSick** [comment]
Is there plan for swagger gen in v4?

**Comment by kamilmysliwiec** [comment]
@KatSick, yes, the swagger module is done

**Comment by BruceHem** [comment]
this will be a major release, will older releases will be affected?

**Comment by d3viant0ne** [comment]
@BruceHem - The major version has been incremented in the `package.json`

**Comment by rhyek** [comment]
Hello! Is there currently a date scheduled for this release?

**Comment by d3viant0ne** [comment]
@kamilmysliwiec is aiming for the end of the month ( most news like this is either in gitter or on twitter )

**Comment by BruceHem** [comment]
@d3viant0ne we are desperate 😄

**Comment by cojack** [comment]
@kamilmysliwiec bump

**Comment by kamilmysliwiec** [comment]
Hey :slightly_smiling_face: v4 is done, I just need to finish the documentation (50% remained).

**Comment by BruceHem** [comment]
Sweet

**Comment by d3viant0ne** [comment]
@kamilmysliwiec - Still probably wouldn't hurt to push out a beta build off the branch and get some of us beating on it before it makes master.

**Comment by bradleyayers** [comment]
@kamilmysliwiec is there anything people can do to help with the documentation?

**Comment by cojack** [comment]
Does examples works?

What I have done:
* checkout to release/v4
* npm i
* rm node_modules/@nestjs
* ln -s ~/Projects/private/nest/src node_modules/@nestjs
* npm start

```
╭─cojack@cojack-laptop  ~/Projects/private/nest/examples/01-cats-app  ‹release/v4*› 
╰─$ npm start

> nest-typescript-starter@1.0.0 start /home/cojack/Projects/private/nest/examples/01-cats-app
> node index.js


/home/cojack/Projects/private/nest/examples/01-cats-app/node_modules/ts-node/src/index.ts:307
        throw new TSError(formatDiagnostics(diagnosticList, cwd, ts, lineOffset))
              ^
TSError: ⨯ Unable to compile TypeScript
../../src/core/scanner.ts (5,48): Module '"/home/cojack/Projects/private/nest/node_modules/@nestjs/common/constants"' has no exported member 'GUARDS_METADATA'. (2305)
../../src/core/scanner.ts (5,65): Module '"/home/cojack/Projects/private/nest/node_modules/@nestjs/common/constants"' has no exported member 'INTERCEPTORS_METADATA'. (2305)
    at getOutput (/home/cojack/Projects/private/nest/examples/01-cats-app/node_modules/ts-node/src/index.ts:307:15)
    at /home/cojack/Projects/private/nest/examples/01-cats-app/node_modules/ts-node/src/index.ts:336:16
    at Object.compile (/home/cojack/Projects/private/nest/examples/01-cats-app/node_modules/ts-node/src/index.ts:498:11)
    at Module.m._compile (/home/cojack/Projects/private/nest/examples/01-cats-app/node_modules/ts-node/src/index.ts:392:43)
    at Module._extensions..js (module.js:580:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/home/cojack/Projects/private/nest/examples/01-cats-app/node_modules/ts-node/src/index.ts:395:12)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Module.require (module.js:513:17)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! nest-typescript-starter@1.0.0 start: `node index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the nest-typescript-starter@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/cojack/.npm/_logs/2017-09-12T01_39_01_488Z-debug.log```

**Comment by nevech** [comment]
@kamilmysliwiec Thank you for your work! What about swagger? I can't find info about it in documentation.

**Pull request merged**